src/solver/opamSolver.ml

```diff
@@ -772,19 +772,27 @@ let dump_universe universe oc =
 
 let filter_solution filter t =
   let t = OpamCudf.ActionGraph.copy t in
-  let rec rm iter_deps v =
-    if OpamCudf.ActionGraph.mem_vertex t v then (
-      iter_deps (rm iter_deps) t v;
+  let rec rm ?(check=fun _ -> true) iter_deps v =
+    if OpamCudf.ActionGraph.mem_vertex t v && check v then (
+      iter_deps (rm ~check iter_deps) t v;
       OpamCudf.ActionGraph.remove_vertex t v
     ) in
   OpamCudf.ActionGraph.iter_vertex
-    (function
-      | `Remove nv as a when not (filter (OpamCudf.cudf2opam nv)) ->
-        rm OpamCudf.ActionGraph.iter_pred a
-      | (`Install nv | `Change (_,_,nv)) as a
-        when not (filter (OpamCudf.cudf2opam nv)) ->
-        rm OpamCudf.ActionGraph.iter_succ a
-      | _ -> ())
+    (fun a ->
+       match a with
+       | `Remove nv as a when not (filter (OpamCudf.cudf2opam nv)) ->
+         rm OpamCudf.ActionGraph.iter_pred a
+       | (`Install nv | `Change (_,_,nv)) as a
+         when not (filter (OpamCudf.cudf2opam nv)) ->
+         (* If a predecessor is a remove, it is a reinstall fragmented *)
+         let check v =
+           List.exists (function
+               | `Remove nv' -> nv = nv'
```

Comment:
In my opinion polymorphic equality should be avoided whenever possible to avoid nasty bugs in the future. Could you use the appropriate `equal` function (or set the type, e.g. ``(nv : [`A | `B | `C]) = nv’``)